python/adjoint/optimization_problem.py

```diff
@@ -229,12 +230,30 @@ def prepare_adjoint_run(self,objective_idx):
             dJ = jacobian(self.objective_functions[objective_idx],mi)(*self.results_list) # get gradient of objective w.r.t. monitor
             if np.any(dJ):
                 self.adjoint_sources += m.place_adjoint_source(dJ) # place the appropriate adjoint sources
-        
+
         # Reset the fields
         self.sim.reset_meep()
-        
+
+
         # Update the sources
-        self.sim.change_sources(self.adjoint_sources)
+        #self.sim.change_sources(self.adjoint_sources)
+
+        if self.adjoint_sources[0] == 'src_vol':
+            self.sim.change_sources([])
+            self.sim.init_sim()
+            for (lincoeff, basisfunc, srcdata) in self.adjoint_sources[1:]:
+                (num_basis, num_pts) = lincoeff.shape
+                for basis_i in range(num_basis):
+                    amps_arr_vec = mp.ComplexVector(num_pts)
+                    for i in range(num_pts):
+                        amps_arr_vec[i] = lincoeff[basis_i, i]
+                    self.sim.fields.add_srcdata(srcdata.fc, srcdata.near_fd_comp, basisfunc[basis_i].swigobj, num_pts, srcdata.idx_arr, amps_arr_vec)
```

Comment:
Why aren't we just creating a volume source in the FarCoefficients class? This routine should simply update the sources and not have to do any "postprocessing" of the sources.

Reply:
You might need to create a new python source that does all of the c++ interfaces once the simulation object is initialized (that looks like what you are trying to hack here). Maybe look at how the other sources are defined.

Reply:
For example, after you create a new class, you can just add an extra condition here that does all the hackery:

https://github.com/NanoComp/meep/blob/859d2198e43c9d3bec00858ceaae87aaa696d4f1/python/simulation.py#L2288-L2294